### PR TITLE
Lexus 2017 IS300H Support

### PR DESCRIPTION
**Pull request by nelsonjchen**
Based on @eFiniLan's https://github.com/commaai/openpilot/pull/536 which got incompletely merged in. Oddly enough, before https://github.com/commaai/openpilot/pull/20789, the README said it was supported. 

Anyway, this PR will be an ongoing WIP to refurbish that PR to actually support the Lexus 2017 IS300*H*.

**Checklist**
- [x] added to README
- [x] test route added to [test_routes.py](../../selfdrive/test/test_routes.py)
- [X] route with openpilot: d72a6d0d7af05505|2021-05-16--11-54-24 (this PR) / d72a6d0d7af05505|2021-05-14--14-39-40 (hacked older branch based on release2/Added to CI Bucket)
- [X] route with stock system: d72a6d0d7af05505|2021-05-17--09-45-01

**Port Specific modernization TODOs**
- [x] Figure out GAS_PRESSED equivalent
- [x] A lot more [checks](https://github.com/commaai/openpilot/blob/4781598e1228b3f8707afa7237864f4cb4df9d71/selfdrive/car/toyota/carstate.py#L139-L152) were added last week. This will need to be re-tested.
- [x] Verify/Modify that Panda safety still works on *both* pedals since the addresses are different.

### CODE REVIEW

File: selfdrive/car/toyota/values.py [hunk 1]
```diff
@@ -941,6 +943,26 @@ class CAR:
       b'8646F5301200\x00\x00\x00\x00',
     ],
   },
+  CAR.LEXUS_ISH: {
```
Comment (adeebshihadeh) [inline selfdrive/car/toyota/values.py:946]: add to the ignored fingerprints list

**Comment by adeebshihadeh** [comment]
The car model unit tests will cover the new checks once you get us a route to upload to the CI bucket. The test can also be run locally for development if you want.

**Comment by nelsonjchen** [comment]
I believe the current CI/CD fails due to the new checks introduced in https://github.com/commaai/openpilot/pull/20742. It'll take me some time to work with the owner to get the PR compliant.

**Comment by nelsonjchen** [comment]
Added a checkbox to check safety since the addresses have shifted and @eFiniLan's memory is a bit fuzzy on the whole disable on gas thing.

**Comment by nelsonjchen** [comment]
Expressed the lack of `GAS_PEDAL` in #toyota-lexus: https://discord.com/channels/469524606043160576/524327905937850394/840117194120757248

**Comment by nelsonjchen** [comment]
@bisra2 reports gas pedal still disengages though.

**Comment by nelsonjchen** [comment]
I haven't forgotten about this. I've just been busy with other stuff. 

[The changes to `toyota/carstate.py` seem to have stabilized](https://github.com/commaai/openpilot/blob/533bc30c2e4ecac5e022f2ba4f3919842dfa0a1a/selfdrive/car/toyota/carstate.py#L111-L136) with the last change being about 20 days ago.  Now seems like a good time to rebase.

@adeebshihadeh , could you copy @bisra2's `d72a6d0d7af05505|2021-05-14--14-39-40` to the CI test bucket so we can develop the needed changes with unit tests?

**Comment by adeebshihadeh** [comment]
Added to the CI bucket.

**Comment by nelsonjchen** [comment]
I haven't tested this yet with @bisra2 on the real vehicle but I've moved the engine FW that was added recently to `LEXUS_IS` to `LEXUS_ISH`. @bisra2's hacked version of `release2` still called it a LEXUS IS 300 2018.

**Comment by nelsonjchen** [comment]
> @bisra2 reports gas pedal still disengages though.

Apparently the codebase uses `GAS_RELEASED` from `PCM_CRUISE` for this and isn't dependent on a bit in `GAS_PEDAL`.

**Comment by nelsonjchen** [comment]
@bisra2 says this branch still works on his IS300H.

**Comment by nelsonjchen** [comment]
I'll still be working with @bisra2 to get the stock route as needed in the checklist.

File: selfdrive/car/toyota/interface.py [hunk 2]
```diff
@@ -258,7 +258,7 @@ def get_params(candidate, fingerprint=gen_empty_fingerprint(), car_fw=[]):  # py
       ret.lateralTuning.pid.kpV, ret.lateralTuning.pid.kiV = [[0.19], [0.02]]
       ret.lateralTuning.pid.kf = 0.00007818594
 
-    elif candidate == CAR.LEXUS_IS:
+    elif candidate in [CAR.LEXUS_IS, CAR.LEXUS_ISH]:
       stop_and_go = False
       ret.safetyParam = 77
```
Comment (eFiniLan) [inline selfdrive/car/toyota/interface.py:263]: @nelsonjchen perhaps push it with 100 or 130, I personally ran 130 and it can take some serious turns.
https://youtu.be/vxq5apbCoq4

**Comment by nelsonjchen** [comment]
Added "d72a6d0d7af05505|2021-05-17--09-45-01" route to description for route with stock system.

File: selfdrive/car/toyota/interface.py (hunk 2 above)
Comment (nelsonjchen) [reply selfdrive/car/toyota/interface.py:263]: I am not familiar with tuning and this value doesn't go that high for any other Toyota. If @bisra2 is interested, maybe he can run a fork with that but I'm going to leave it as-is for this PR.

**Comment by nelsonjchen** [comment]
@bisra2 doesn't have Comma Prime and I don't know what to do about the routes of `d72a6d0d7af05505|2021-05-16--11-54-24` and ` d72a6d0d7af05505|2021-05-17--09-45-01` eventually getting deleted.

File: selfdrive/car/toyota/values.py [hunk 3]
```diff
@@ -296,8 +298,19 @@ class CAR:
 }
 
 # Don't use theses fingerprints for fingerprinting, they are still needed for ECU detection
-IGNORED_FINGERPRINTS = [CAR.RAV4H_TSS2, CAR.HIGHLANDERH_TSS2, CAR.LEXUS_RXH_TSS2, CAR.PRIUS_TSS2,
-                        CAR.LEXUS_NX, CAR.LEXUS_NX_TSS2, CAR.CAMRY_TSS2, CAR.CAMRYH_TSS2, CAR.LEXUS_ESH, CAR.MIRAI]
+IGNORED_FINGERPRINTS = [
```
Comment (adeebshihadeh) [inline selfdrive/car/toyota/values.py:301]: revert the style change, just add to the list

File: README.md [hunk 4]
```diff
@@ -90,6 +90,7 @@ Supported Cars
 | Lexus     | ES Hybrid 2017-18             | LSS               | Stock<sup>3</sup>| 0mph               | 0mph              |
 | Lexus     | ES Hybrid 2019-21             | All               | openpilot        | 0mph               | 0mph              |
 | Lexus     | IS 2017-2019                  | All               | Stock            | 22mph              | 0mph              |
+| Lexus     | IS Hybrid 2017-2019           | All               | Stock            | 22mph              | 0mph              |
```
Comment (adeebshihadeh) [inline README.md:93]: are all these years confirmed working?

File: selfdrive/car/toyota/carcontroller.py [hunk 5]
```diff
@@ -112,7 +112,7 @@ def update(self, enabled, CS, frame, actuators, pcm_cancel_cmd, hud_alert,
       lead = lead or CS.out.vEgo < 12.    # at low speed we always assume the lead is present do ACC can be engaged
 
       # Lexus IS uses a different cancellation message
-      if pcm_cancel_cmd and CS.CP.carFingerprint == CAR.LEXUS_IS:
+      if pcm_cancel_cmd and CS.CP.carFingerprint in [CAR.LEXUS_IS, CAR.LEXUS_ISH]:
```
Comment (adeebshihadeh) [inline selfdrive/car/toyota/carcontroller.py:115]: is this confirmed to work?

File: selfdrive/test/test_routes.py [hunk 6]
```diff
@@ -329,6 +329,11 @@
     'enableCamera': True,
     'enableDsu': True,
   },
+  "d72a6d0d7af05505|2021-05-14--14-39-40": {
+    'carFingerprint': TOYOTA.LEXUS_ISH,
+    'enableCamera': True,
+    'enableDsu': False,
```
Comment (adeebshihadeh) [inline selfdrive/test/test_routes.py:335]: remove these two lines, they're unused now

File: selfdrive/car/toyota/values.py [hunk 7]
```diff
@@ -925,7 +938,6 @@ class CAR:
       b'\x018966353Q4000\x00\x00\x00\x00',
     ],
     (Ecu.engine, 0x7e0, None): [
-      b'\x0232480000\x00\x00\x00\x00\x00\x00\x00\x00A4701000\x00\x00\x00\x00\x00\x00\x00\x00',
```
Comment (adeebshihadeh) [inline selfdrive/car/toyota/values.py:928]: why was this removed?

Comment (nelsonjchen) [reply selfdrive/car/toyota/values.py:928]: This was the exact engine fw on @bisra2's ISH. This firmware was added in https://github.com/commaai/openpilot/pull/20850 as part of a batch. It seemed extremely unlikely that the gas-only model has the same engine firmware as the hybrid and I think this was an error.

File: README.md (hunk 4 above)
Comment (nelsonjchen) [reply README.md:93]: No, I'll move it back to 2017 for now or whatever @bisra2's ISH is.

File: selfdrive/car/toyota/values.py (hunk 3 above)
Comment (nelsonjchen) [reply selfdrive/car/toyota/values.py:301]: Sorry, will do.

File: selfdrive/car/toyota/carcontroller.py (hunk 5 above)
Comment (nelsonjchen) [reply selfdrive/car/toyota/carcontroller.py:115]: This is usually used for the DM check failing out right?

File: selfdrive/car/toyota/values.py (hunk 7 above)
Comment (nelsonjchen) [reply selfdrive/car/toyota/values.py:928]: FWIW, I think we might see a similar situation with the XX40 Avalons and XX50 Avalons. I want to send a PR in to separate those outs later but firmwares from users who have hacked their installations to get XX50 Avalons working by changing the DBCs used are still identifying as the presumably XX40 Avalon. 

I think something similar happened here for the ISH since @bisra2 was running a hacked version that was still identifying as an IS before I made this proper separated-out branch.

File: selfdrive/car/toyota/carcontroller.py (hunk 5 above)
Comment (adeebshihadeh) [reply selfdrive/car/toyota/carcontroller.py:115]: this is used any time openpilot disengages itself

Comment (nelsonjchen) [reply selfdrive/car/toyota/carcontroller.py:115]: I'll work with @bisra2 to test this with the seatbelt unbuckled disengagement. 

I've actually never had OP disengage on me ever except due to kernel panics or some software/hardware issue so I think I'll try it myself on my own car first before guiding @bisra2 on this way.

File: README.md (hunk 4 above)
Comment (nelsonjchen) [reply README.md:93]: @bisra2 says his is the 2018 so I will edit the README to just 2018.

File: selfdrive/car/toyota/carcontroller.py (hunk 5 above)
Comment (adeebshihadeh) [reply selfdrive/car/toyota/carcontroller.py:115]: open an issue if you're hitting kernel panics, that shouldn't happen

Comment (nelsonjchen) [reply selfdrive/car/toyota/carcontroller.py:115]: > open an issue if you're hitting kernel panics, that shouldn't happen

It happened a long time ago on mine, and many months ago. I'll keep that in mind to do if it happens again.

Comment (nelsonjchen) [reply selfdrive/car/toyota/carcontroller.py:115]: @bisra2, I don't think the cruise control cancellation command is working here or at least openpilot isn't disengaging. When the seatbelt is unbuckled, there's no period of cruise control being disengaged afterwards. 

@eFiniLan , it's been a long time, but do you remember testing this functionality on your IS300H when you had it? 

I think we might need to figure out an alternate working command here. 

@bisra2's IS300H:

It goes Green, Red, and Green. The `CRUISE_ACTIVE` signal never goes low during the seatbelt unbuckling.

image

image


My 2020 Corolla:

It goes Green, Red, Blue, and Green as I turn back on openpilot after unbuckling the seatbelt. The `CRUISE_ACTIVE` signal goes low during the seatbelt unbuckling.

image

image

Comment (eFiniLan) [reply selfdrive/car/toyota/carcontroller.py:115]: @nelsonjchen as far as I remembered that part never worked, OP itself disengaged but stock ACC remains.

Comment (nelsonjchen) [reply selfdrive/car/toyota/carcontroller.py:115]: I've asked ash11111111#7507 about the OP disengage on gas on Discord who has a US/CAN IS300 but they are busy fighting some other hardware issue with their Comma Two at the moment to respond. ash11111111#7507 did say the stock system doesn't disengage on gas but we all know that.

Comment (adeebshihadeh) [reply selfdrive/car/toyota/carcontroller.py:115]: I can confirm the IS300 disengages on gas and works with the current cancellation message.

Comment (adeebshihadeh) [reply selfdrive/car/toyota/carcontroller.py:115]: The seatbelt unlatched event is a soft disable, meaning it doesn't disengage immediately. Gas pedal disengage would be the most direct way to test the cancellation. This is a blocker for merging.

Comment (nelsonjchen) [reply selfdrive/car/toyota/carcontroller.py:115]: Ok, I'll put this back to draft for now. In the meantime, I will repeat what I received from @bisra2 after asking "To be clear: Pressing the gas pedal disengages openpilot, but the automatic cruise control continues to be enabled?":


Test 1: 
Openpilot active, unbuckling the seatbelt: The stock automatic cruise control continues to be enabled. The car slows down and accelerates on its own depending on the traffic but no more openpilot
Test2:
Openpilot activated, I press the gas pedal.:The stock automatic cruise control continues to be enabled. The car slows down and accelerates on its own depending on the traffic but no more openpilot
So the answer to your question is: Yes
To summarize, pressing the gas pedal or unbuckling the seat belt causes openpilot to disengage, but the stock safety sense continues to operate in both cases.

**Comment by nelsonjchen** [comment]
It's unlikely I'll be able to find the correct CAN command to disable ACC remotely from Arizona with someone in France. I'll close this PR for now and hopefully whoever tries to produce a mergable pull request in the future has better access to the car can base their work on this.
